Index: iputils/ping.c
===================================================================
--- iputils.orig/ping.c	2013-12-29 23:23:02.000000000 -0800
+++ iputils/ping.c	2013-12-29 23:24:38.000000000 -0800
@@ -1243,7 +1243,13 @@
 pr_addr(__u32 addr)
 {
 	struct hostent *hp;
-	static char buf[4096];
+	static char buf[4096] = "";
+	static __u32 last_addr = 0;
+
+	if (*buf && addr == last_addr)
+		return(buf);
+
+	last_addr = addr;
 
 	in_pr_addr = !setjmp(pr_addr_jmp);
 
Index: iputils/ping6.c
===================================================================
--- iputils.orig/ping6.c	2013-12-29 23:23:02.000000000 -0800
+++ iputils/ping6.c	2013-12-29 23:24:38.000000000 -0800
@@ -1794,8 +1794,9 @@
  */
 char * pr_addr(struct in6_addr *addr)
 {
-	struct hostent *hp = NULL;
+	static struct hostent *hp = NULL;
 	static char *s;
+	static struct in6_addr last_addr;
 
 #ifdef USE_IDN
 	free(s);
@@ -1803,8 +1804,11 @@
 
 	in_pr_addr = !setjmp(pr_addr_jmp);
 
-	if (!(exiting || options&F_NUMERIC))
+	if (!(hp && memcmp(addr, &last_addr, sizeof(struct in6_addr))) &&
+	    !(exiting || options&F_NUMERIC)) {
 		hp = gethostbyaddr((__u8*)addr, sizeof(struct in6_addr), AF_INET6);
+		memcpy(&last_addr, addr, sizeof(struct in6_addr));
+	}
 
 	in_pr_addr = 0;
 
