Index: iputils/doc/Makefile
===================================================================
--- iputils.orig/doc/Makefile	2014-02-01 12:27:29.164716082 -0800
+++ iputils/doc/Makefile	2014-02-01 12:27:48.864813759 -0800
@@ -6,7 +6,7 @@
 
 html: $(HTMLFILES) iputils.html
 
-man: $(MANFILES)
+man: $(MANFILES) fix_sgml2man
 
 # docbook scripts are incredibly dirty in the sense that they leak
 # lots of some strange temporary junk directories and files.
@@ -33,6 +33,9 @@
 	@set -e; cd tmp.db2man; nsgmls ../$< | sgmlspl ../docbook2man-spec.pl ;	mv $@ ..
 	@-rm -rf tmp.db2man
 
+fix_sgml2man:
+	@sed -i -e 's!\\fB\\fIdestination\\fB\\fR \[\\fB/\\fIport\\fB\\fR\]!\\fB\\fIdestination\\fB\\fR[\\fB/\\fIport\\fB\\fR]!g' tracepath.8
+
 clean:
 	@rm -rf $(MANFILES) $(HTMLFILES) iputils.html tmp.db2html tmp.db2man
 
Index: iputils/doc/tracepath.sgml
===================================================================
--- iputils.orig/doc/tracepath.sgml	2014-02-01 12:27:28.508712829 -0800
+++ iputils/doc/tracepath.sgml	2014-02-01 12:29:16.509249249 -0800
@@ -46,21 +46,21 @@
 <variablelist>
 
  <varlistentry>
-  <term><option/-n/</term>
+  <term><option>-n</option></term>
   <listitem><para>
 Print primarily IP addresses numerically.
   </para></listitem>
  </varlistentry>
 
  <varlistentry>
-  <term><option/-b/</term>
+  <term><option>-b</option></term>
   <listitem><para>
 Print both of host names and IP addresses.
   </para></listitem>
  </varlistentry>
 
  <varlistentry>
-  <term><option/-l/</term>
+  <term><option>-l</option></term>
   <listitem><para>
 Sets the initial packet length to <replaceable/pktlen/ instead of
 65535 for <command/tracepath/ or 128000 for <command/tracepath6/.
@@ -68,7 +68,7 @@
  </varlistentry>
 
  <varlistentry>
-  <term><option/-p/</term>
+  <term><option>-p</option></term>
   <listitem><para>
 Sets the initial destination port to use.
   </para></listitem>
