Cherry-picked from upstream revision 608419a7
Index: iputils/doc/tracepath.sgml
===================================================================
--- iputils.orig/doc/tracepath.sgml	2014-02-01 12:56:47.641435899 -0800
+++ iputils/doc/tracepath.sgml	2014-02-01 12:57:00.373499033 -0800
@@ -18,6 +18,7 @@
 <arg choice="opt">-n</arg>
 <arg choice="opt">-b</arg>
 <arg choice="opt">-l <replaceable/pktlen/</arg>
+<arg choice="opt">-m <replaceable/max_hops/</arg>
 <arg choice="opt">-p <replaceable/port/</arg>
 <arg choice="req"><replaceable/destination/</arg>
 </cmdsynopsis>
@@ -67,6 +68,14 @@
   </para></listitem>
  </varlistentry>
 
+ <varlistentry>
+  <term><option>-m</option>/term>
+  <listitem><para>
+Set maximum hops (or maximum TTLs) to <replaceable/max_hops/
+instead of 30.
+  </para></listitem>
+ </varlistentry>
+
  <varlistentry>
   <term><option>-p</option></term>
   <listitem><para>
Index: iputils/tracepath.c
===================================================================
--- iputils.orig/tracepath.c	2014-02-01 12:56:44.709421358 -0800
+++ iputils/tracepath.c	2014-02-01 12:56:54.349469161 -0800
@@ -32,6 +32,9 @@
 #define IP_PMTUDISC_PROBE	3
 #endif
 
+#define MAX_HOPS_LIMIT		255
+#define MAX_HOPS_DEFAULT	30
+
 struct hhistory
 {
 	int	hops;
@@ -43,6 +46,7 @@
 
 struct sockaddr_in target;
 __u16 base_port;
+int max_hops = MAX_HOPS_DEFAULT;
 
 const int overhead = 28;
 int mtu = 65535;
@@ -315,7 +319,7 @@
 	setlocale(LC_ALL, "");
 #endif
 
-	while ((ch = getopt(argc, argv, "nbh?l:p:")) != EOF) {
+	while ((ch = getopt(argc, argv, "nbh?l:m:p:")) != EOF) {
 		switch(ch) {
 		case 'n':
 			no_resolve = 1;
@@ -330,6 +334,14 @@
 				exit(1);
 			}
 			break;
+		case 'm':
+			max_hops = atoi(optarg);
+			if (max_hops < 0 || max_hops > MAX_HOPS_LIMIT) {
+				fprintf(stderr,
+					"Error: max hops must be 0 .. %d (inclusive).\n",
+					MAX_HOPS_LIMIT);
+			}
+			break;
 		case 'p':
 			base_port = atoi(optarg);
 			break;
@@ -405,7 +417,7 @@
 		exit(1);
 	}
 
-	for (ttl=1; ttl<32; ttl++) {
+	for (ttl = 1; ttl <= max_hops; ttl++) {
 		int res;
 		int i;
 
Index: iputils/tracepath6.c
===================================================================
--- iputils.orig/tracepath6.c	2014-02-01 12:49:14.311187959 -0800
+++ iputils/tracepath6.c	2014-02-01 12:56:54.349469161 -0800
@@ -42,6 +42,9 @@
 #define IPV6_PMTUDISC_DO	3
 #endif
 
+#define MAX_HOPS_LIMIT		255
+#define MAX_HOPS_DEFAULT	30
+
 struct hhistory
 {
 	int	hops;
@@ -55,6 +58,7 @@
 struct sockaddr_storage target;
 socklen_t targetlen;
 __u16 base_port;
+int max_hops = MAX_HOPS_DEFAULT;
 
 int overhead;
 int mtu;
@@ -387,7 +391,7 @@
 	setlocale(LC_ALL, "");
 #endif
 
-	while ((ch = getopt(argc, argv, "nbh?l:p:")) != EOF) {
+	while ((ch = getopt(argc, argv, "nbh?l:m:p:")) != EOF) {
 		switch(ch) {
 		case 'n':
 			no_resolve = 1;
@@ -398,6 +402,14 @@
 		case 'l':
 			mtu = atoi(optarg);
 			break;
+		case 'm':
+			max_hops = atoi(optarg);
+			if (max_hops < 0 || max_hops > MAX_HOPS_LIMIT) {
+				fprintf(stderr,
+					"Error: max hops must be 0 .. %d (inclusive).\n",
+					MAX_HOPS_LIMIT);
+			}
+			break;
 		case 'p':
 			base_port = atoi(optarg);
 			break;
@@ -523,7 +535,7 @@
 		exit(1);
 	}
 
-	for (ttl=1; ttl<32; ttl++) {
+	for (ttl = 1; ttl <= max_hops; ttl++) {
 		int res;
 		int i;
 
